perfaware/sim86/shared/Sim8086.cpp

```diff
@@ -30,7 +28,69 @@ static char const* RegisterNames[][3] =
     {"flags", "flags", "flags"}
 };
 
-static void SimulateInstruction(const instruction& Instruction)
+static char const* FlagNames = { "ODITSZAPC" };
+enum Flags
+{
+    Flag_OF,
+    Flag_DF,
+    Flag_IF,
+    Flag_TF,
+    Flag_SF,
+    Flag_ZF,
+    Flag_AF,
+    Flag_PF,
+    Flag_CF,
+
+    Flag_number
+};
+
+static void SetFlags(u16 Result, bool* Flags)
+{
+    Flags[Flag_SF] = Result & 0x8000;
+    Flags[Flag_ZF] = (Result == 0);
+}
+
+static void PrintFlags(bool* Flags)
+{
+    std::string OutputBuffer = "Flags: ";
+    for (int i = 0; i < Flag_number; i++)
```

Comment:
I set you a trap to be able to bring up a topic later hehe. Because Flag_number is now an enum value, by default it is unsigned. So in this for loop, you are comparing a signed value (`i`) to an unsigned one (`Flag_number`). Here you don't have a bug, it works. However, this is usually something to be avoided. The signed type gets converted to unsigned before doing the comparison. It's a very common source of bugs, because if `i` had a negative value, the result could have been wrong. So it's good to be aware of such potential pitfalls.

In this specific case, the simple "fix" is to explicitly use an unsigned type for `i` too. Get in the habit of paying attention to situations like this, because it can be subtle to find if you have a bug!

Reply:
Hah! Nice trap. I'll be more careful
079c96f6cdc4673bd76002403380fcb313b41829

Reply:
You used u8, but in a situation like this I would just have used `unsigned int` without specifying an explicit size.

Usually I only specify an explicit size when I have a good reason to want something to be of a specific size. When I just want an integer and I don't care about its size, I use `int` or `unsigned int`. This way, when I use a specific size, the people who read the code know that there's a good reason for the size.

Here there is no good reason to use u8. Usually you use that when you want to represent a byte, or situations like that. For an index, it's not useful. If anything, I would have used u64, this way you minimize chances of overflow.

Also, maybe you did this thinking "hey, it will save space, it's less bits". But it doesn't work like that, because you're using the 8 lower bits of a register, so nobody can use the rest of the register anyway.

Reply:
Or you can also use `size_t`. I don't know if you have learned about it yet, but it's typically used to represent indices, sizes of arrays, things like that, because it's guaranteed to be large enough to hold whatever index you need to access things in an array stored in memory.

Reply:
> But it doesn't work like that, because you're using the 8 lower bits of a register, so nobody can use the rest of the register anyway.

I'll remember that, thanks!
96ced2abff1cee7ddb8028c67c8081d072442884

Reply:
Missed the size_t suggestion, I like the idea
08643fe1b8d34c069a04a98acce49edadf672aea